wrap.man
.TH wrap 1 "15 Dec 1998" "Version 1.3"
.SH NAME
wrap
\- filter to word-wrap text
.SH SYNOPSIS
\fBwrap\fP [\fB-w\fP \fIwidth\fP] [\fB--version --help\fP]
.SH DESCRIPTION
.PP
This program reads text from standard input, word-wraps it, and writes
the result to standard output. The text is formatted to fit a
specified number of columns. Extraneous whitespace is removed and
lines are only broken on word boundaries.
.SH OPTIONS
.TP 5
.B --help
Display a command synopsis and exit.
.TP 5
.B --version
Display the program's version information and exit.
.TP 5
\fB-w\fP \fIwidth\fP
Specify the maximum number of columns \fIwidth\fP that the text may
span. The minimum value of \fIwidth\fP is 10.
.SH SEE ALSO
.BR less (1),
.BR fold (1)
.SH NOTES
If \fBwrap\fP encounters a word that is longer than the specified
column width, the word is broken across two or more lines to fit
within that number of columns.
.SH AUTHOR
Mark Lindner



ChangeLog
2017-10-27  Mark Lindner  <markl@avalon>

	* *.l - fixed comment which caused flex error
	* <multiple> - cleaned up and updated build files

2007-12-29  Mark Lindner  <markl@neuromancer>

	* talkfilters.c, talkfilters.h, Makefile.am, Makefile.in, config.h,
	  - MinGW & VC++ build fixes
	* config.guess, config.sub, ltmain.sh, aclocal.m4, libtool
	  - updated to newest

2006-11-21  Mark Lindner  <markl@neuromancer>

	* pirate.l - minor improvements
	* dubya.l - minor improvements
	* common.h - use VERSION instead of hardcoded version number

2006-10-27  Mark Lindner  <markl@neuromancer>

	* common.c - fix compiler warning
	* *.sln, *.vcproj, win32/unistd.h - add Visual C++ Express 2005
	  build files
	* talkfilters.c - update version number
	* common.h - use C99 varargs macro notation; added some #defines for
	  win32
	* pirate.l - added rule
	* talkfilters.h - added extern "C" guard

2006-09-18  Mark Lindner  <markl@neuromancer>

	* common.c - fix compiler warnings
	* talkfilters.h, talkfilters.c, configure.in, Makefile.am
	  - changes for DLL creation

2006-04-17  Mark Lindner  <markl@neuromancer>

	* funetak.l - replaced random() with gtf_random(); eliminates build
	  error in mingw

2006-04-15  Mark Lindner  <markl@neuromancer>

	* <multiple> - added support for pkg-config

2005-11-21  Mark Lindner  <markl@neuromancer>

	* gaim_plugin/talkfilters.c, Makefile.am - Removed the GAIM plugin,
	  as it's now being maintained elsewhere.
	* common.c, talkfilters.c, common.h, configure.in - Build fixes.
	* talkfilters.spec - added new file
	The above all contributed by Peter Lawler

2005-01-10  Mark Lindner  <markl@neuromancer>

	* gaim_plugin/talkfilters.c - Updated to work with GAIM 1.1.0.

2004-06-28  Mark Lindner  <markl@neuromancer>

	* gaim_plugin/talkfilters.c - Incorporated bug fix submitted by a user.
	  The bug caused the IM to only be filtered locally (not sent in
	  filtered form to the other end). This must be a regression when
	  moving to GAIM 0.78/0.79.

2004-06-26  Mark Lindner  <markl@neuromancer>

	* gaim_plugin/* - Updated plugin to work with GAIM 0.79, and clarified
	  installation instructions.

2004-01-06  Mark Lindner  <markl@neuromancer>

	* funetak.l - Fixed a capitalization problem.
	* austro.l - Added missing rules, whose absence caused garbled output
	  from filter function.
	* gaim_plugin/* - New plugin for GAIM.

2003-12-02  Mark Lindner  <markl@neuromancer>

	* *.l, *.c, *.h - Added contributed austro and dubya filters.

2003-07-22  Mark Lindner  <markl@neuromancer>

	* *.l - Added rules to leave HTML tags intact. Do you see where
          I'm going with this? ;-)
	* kraut.l - Minor fixes and corrections (submitted by Thomas M. Ott)
	* talkfilters.man - Documentation updates.
	* configure.in - Rewrote `flex' check.

2003-04-29  Mark Lindner  <markl@neuromancer>

	* jive.l - Added a few more rules.

2003-02-13  Mark Lindner  <markl@dystance.net>

	* *.l - Made a fix so that compilation doesn't fail with older
	versions of gcc: moved declaration of _yybuf so that it's before the
	first statement in the filter function.

2003-02-11  Mark Lindner  <markl@dystance.net>

	* <multiple> - Complete overhaul of package; added automake/libtool
	support, and produce not only executables but also a library.

Tue Jul  4 21:33:08 2000  Mark A. Lindner <markl@dystance.net>

	* Makefile.in: Modified `dist' target and added `info' target.

	* talkfilters.texi: Texinfo documentation added.

Thu Dec 31 17:09:45 1998  Joel N. Weber II  <devnull@melange.gnu.org>

	* configure.in: Rewrote the check for lex to just check for flex
	and bail if it isn't found.

	* src/talkfilters-1.0/Makefile.in (LIBS): Hardcode -lfl, since we
	always use flex.



wrap.c
/* GNU TalkFilters
   Copyright (C) 1998-2003 Free Software Foundation, Inc.

   This file is part of GNU Talkfilters

   GNU Talkfilters is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   GNU Talkfilters is distributed in the hope that it will be amusing,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with GNU Talkfilters; see the file COPYING.  If not, write to
   the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/* --- System Headers --- */

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>

/* --- Local Headers --- */

/* --- Macros --- */

#define DFL_WIDTH 78
#define MIN_WIDTH 10

#include "getopt.h"

#define HELP_OPTION 0
#define VERSION_OPTION 1

#define USAGE "%s: Usage: %s [-w width] [--help --version]\n"
#define VERSION "GNU wrap v1.3"

static const struct option options[] = { { "help", 0, NULL, 0  },
                                         { "version", 0, NULL, 0 },
                                         { NULL, 0, NULL, 0 } };

/* --- Functions --- */

int main(int argc, char **argv)
{
  char *wbuf = (char *)malloc((DFL_WIDTH + 1) * sizeof(char)), *wp = wbuf,
    prevc = 0, errf = 0;
  int c, wl = 0, ll = 0, width = DFL_WIDTH, option_index;
  extern char *optarg;

  while((c = getopt_long(argc, argv, "w:", options, &option_index)) != EOF)
  {
    switch(c)
    {
      case 'w':
        width = atoi(optarg);
        if(width < MIN_WIDTH)
        {
          fprintf(stderr,
                  "%s: Width must be at least %i (defaulting to %i)\n",
                  *argv, MIN_WIDTH, DFL_WIDTH);
          width = DFL_WIDTH;
        }
        break;

      case 0:
        if(option_index == HELP_OPTION)
          fprintf(stderr, USAGE, *argv, *argv);
        else if(option_index == VERSION_OPTION)
        {
          fputs(VERSION, stderr);
          fputc('\n', stderr);
        }
        exit(EXIT_SUCCESS);
        break;

      default:
        errf++;
    }
  }

  if(errf)
  {
    fprintf(stderr, USAGE, *argv, *argv);
    free(wbuf);
    exit(EXIT_FAILURE);
  }

  /* now do the actual formatting */

  while((c = getchar()) != EOF)
  {
    if(!isspace(c))
    {
      if(wl < width)
      {
        *wp = (char)c;
        wl++, wp++;
      }
      else /* word is full, so output it */
      {
        *wp = 0;
        putchar('\n');
        fputs(wbuf, stdout);
        putchar(c);
        putchar('\n');
        wp = wbuf, ll = wl = 0;
      }
    }
    else /* isspace(c) */
    {
      if(prevc == '\n')
      {
        if(prevc == '\n' && c == '\n')
          ll = 0, putchar('\n'), putchar('\n');
        else if(ll > 0) ll++;
      }
      else if(wl && ((ll + wl + (ll > 0)) <= width))
      {
        *wp = 0;
        if(ll) putchar(' '), ll++;
        fputs(wbuf, stdout);
        ll += wl;
        wl = 0, wp = wbuf;
      }
      else if(wl)
      {
        *wp = 0;
        putchar('\n');
        fputs(wbuf, stdout);
        ll = wl;
        wl = 0, wp = wbuf;
      }
    }
    prevc = c;
  }
  /* if text remains in word buffer, process it */

  if(wl)
  {
    *wp = 0;
    if((ll + wl + (ll > 0)) <= width)
    {
      if(ll) putchar(' ');
    }
    else putchar('\n');
    fputs(wbuf, stdout);
  }
  putchar('\n');
  free(wbuf);
  exit(EXIT_SUCCESS);
}

/* end of source file */
